openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ClassProfiler.java

```diff
@@ -0,0 +1,348 @@
+			re.printStackTrace();
+			Assert.fail("Problem with the reflection of the MBean");
+		}
+
+		// The first time we write to the file, 'append' needs to be false.
+		// From the second time onwards, it needs to be true
+		try {
+			this.envData.writeData(this.mbs, srvRuntimeName, srvOSName, srvLogName, false);
+			int secondsCnt = 0;
+			int writesCnt = 0;
+			while (writesCnt < 30) {
+				// Write out the threadData every 10 seconds
+				if (secondsCnt == 10) {
+					this.classData.writeData(this.mbs, srvClassName, srvCompName, srvRuntimeName, true);
+					secondsCnt = 0;
+					writesCnt++;
+				}
+				this.recordClassStats(srvRuntimeName, srvClassName);
+				Thread.sleep(1000);
+				secondsCnt++;
+			}
+
+			// If the exception is a Connect or Unmarshal Exception, assume the
+			// monitored JVM has finished 
+		} catch (ConnectException ce) {
+			this.closeCSVFile();
+			Message.logOut("Exiting as JVM we are connected to has finished");
+			Assert.fail("Exiting as JVM we are connected to has finished");
+		} catch (UnmarshalException ue) {
+			this.closeCSVFile();
+			Message.logOut("Exiting as JVM we are connected to has finished");
+			Assert.fail("Exiting as JVM we are connected to has finished");
+		} catch (InterruptedException ie) {
+			Message.logOut("The sleeping profiler was interrupted");
+			ie.printStackTrace();
+			Assert.fail("The sleeping profiler was interrupted");
+		} finally {
+			this.closeCSVFile();
+		}
+	}
+
+	private void createCSVFile(String csvFile) {
+		try {
+			// Open the cvs file
+			this.csv = new PrintWriter(new FileWriter(new File(csvFile), false));
+			// write headings to file
+			this.csv.println("Time (seconds), Loaded Class Count, Total Loaded Class Count, Unloaded Class Count");
+		} catch (IOException ie) {
+			Message.logOut("Unable to write to file " + csvFile);
+			ie.printStackTrace();
+			Assert.fail("Unable to write to file " + csvFile);			
+		}
+	}
+
+	private void recordClassStats(RuntimeMXBean rb, ClassLoadingMXBean cb) {
+		// Print out time in 2 d.p.
+		DecimalFormat df = new DecimalFormat("#0.00"); 
+		// Write stats to file
+		this.csv.println(df.format(rb.getUptime() / 1000) + "," + cb.getLoadedClassCount() + ","
+				+ cb.getTotalLoadedClassCount() + "," + cb.getUnloadedClassCount());
+	}
+
+	private void recordClassStats(ObjectName rb, ObjectName cb) {
+		// Print out time in 2 d.p.
+		DecimalFormat df = new DecimalFormat("#0.00");  
+		try {
+			int classCount = (Integer)this.mbs.getAttribute(cb, "LoadedClassCount");
+			long loadedCount = (Long)this.mbs.getAttribute(cb, "TotalLoadedClassCount");
+			long unloadedCount = (Long)this.mbs.getAttribute(cb, "UnloadedClassCount");
+			long upTime = (Long)this.mbs.getAttribute(rb, "Uptime");
+
+			// Write class data to file
+			this.csv.println(df.format(upTime / 1000) + "," + classCount + "," + 
+							loadedCount + "," + unloadedCount);
+		} catch (ConnectException ce) {
+			Message.logOut("Problem with the MBean access");
```

Comment:
Same comment as previous about the logging story applies across many test classes.